### PR TITLE
[SPARK-32653][CORE] Decommissioned host/executor should be considered as inactive in TaskSchedulerImpl

### PR DESCRIPTION
**Pull request by Ngone51**
### What changes were proposed in this pull request?

Add decommissioning status checking for a host or executor while checking it's active or not. And a decommissioned host or executor should be considered as inactive.

### Why are the changes needed?

First of all, this PR is not a correctness bug fix but gives improvement indeed. And the main problem here we want to fix is that a decommissioned host or executor should be considered as inactive.

`TaskSetManager.computeValidLocalityLevels` depends on `TaskSchedulerImpl.isExecutorAlive/hasExecutorsAliveOnHost` to calculate the locality levels. Therefore, the `TaskSetManager` could also get corresponding locality levels of those decommissioned hosts or executors if they're not considered as inactive. However, on the other side,  `CoarseGrainedSchedulerBackend` won't construct the `WorkerOffer` for those decommissioned executors. That also means `TaskSetManager` might never have a chance to launch tasks at certain locality levels but only suffers the unnecessary delay because of delay scheduling. So, this PR helps to reduce this kind of unnecessary delay by making decommissioned host/executor inactive in `TaskSchedulerImpl`.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Added unit tests

### CODE REVIEW

**Comment by Ngone51** [comment]
@cloud-fan @agrawaldevesh @holdenk Please take a look, thanks!

File: core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala [hunk 1]
```diff
@@ -1062,25 +1062,36 @@ private[spark] class TaskSchedulerImpl(
...
 
   def hasExecutorsAliveOnHost(host: String): Boolean = synchronized {
-    hostToExecutors.contains(host)
+      hostToExecutors.get(host)
+        .exists(executors => executors.exists(e => !isExecutorDecommissioned(e)))
   }
 
   def hasHostAliveOnRack(rack: String): Boolean = synchronized {
-    hostsByRack.contains(rack)
+    hostsByRack.get(rack)
```
Comment (agrawaldevesh) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala:1074]: I haven't looked at this code too deeply, but does it make sense it eagerly remove the decommissioned executor from hostToExecutors and recomputeLocality and thus rebuild hostsByRack etc ?

**Review by agrawaldevesh — COMMENTED** [review]
Thanks for fixing this @Ngone51 but the rationale isn't very clear to me yet.

Per my understanding the coarse grained executor wouldn't even assign new tasks to decommissioned executors. Perhaps you can update the PR description with a case where this is warranted ? Like what happens if we don't have this fix ?

In the meanwhile, I will play with the code in this PR to understand it better.

**Review by agrawaldevesh — CHANGES_REQUESTED** [review]
It is entirely possible I am missing something below so please do let me know if there is a bug in my understanding. Thanks for looking into this with eagle eyes and fixing such gaps !.

I went through the code and per my understanding I think this PR is not strictly required from a correctness perspective, but it may buy us some driver efficiency. 

Basically, the TaskSchedulerImpl/TaskSetManager should never see a decommissioned executor in `TaskSchedulerImpl.resourceOffers(...)`, so it should never inadvertently schedule a task on a decommissioned executor -- that is why I think that this PR does not fix a correctness bug.

OTOH, in certain cases when decommissioned executors have some cached/shuffle data on them -- ie if they are a part of the Task.preferredLocations, then `computeValidLocalityLevels` might set a level that cannot be satisfied. This will cause some extra checking in the driver. Example: A shuffle map task T wants to run on host A, ie `pendingTasks.forHost[A].contains(T)`. So a task set containing T might set its locality level to also include NODE_LOCAL. However, the TaskSchedulerImpl will never be shown an executor on A, and thus this locality level will never be satisfied with A. Therefore, this will only result in some extra scheduling cycles on the driver. 

I would please like to better understand the motivation of this PR before proceeding further with the review: Exactly what problem is it trying to solve ? What happens if we don't solve it ? I am marking it as "Request Changes" so that we first understand the problem better.

**Comment by Ngone51** [comment]
Have synced with @agrawaldevesh offline, I've updated the PR description according to his questions.

File: core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala (hunk 1 above)
Comment (Ngone51) [reply core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala:1074]: `hostToExecutors` will also be used to handle executor removal and blacklisting stuff. In other words, a decommissioned executor could also be removed or blacklisted at the same time. So, remove the executor from `hostToExecutors` might also affect other places. I don't want to mix them together to make things complex.

Besides, similar to `executorIdToRunningTaskIds`, a decommissioned executor could also have running tasks on it. So we don't eagerly remove it from `executorIdToRunningTaskIds` too.

File: core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala [hunk 2]
```diff
@@ -1062,25 +1062,36 @@ private[spark] class TaskSchedulerImpl(
   }
 
   def getExecutorsAliveOnHost(host: String): Option[Set[String]] = synchronized {
-    hostToExecutors.get(host).map(_.toSet)
+    hostToExecutors.get(host).map(_.filterNot(isExecutorDecommissioned)).map(_.toSet)
```
[the remark this answers is not in the document]
Comment (Ngone51) [reply core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala:1065]: hmm..I think this more like a problem of definition or something like that. In this PR, what I'm trying to do is to strengthen the definition of the active status for the executor or host. That is, for an active executor or host, it should not be decommissioned at the same time. This function is getting the active host, so we should filter out decommissioned ones. I think this's quite obvious. And this function is used by multiple tests.

**Review by HyukjinKwon — APPROVED** [review]

File: core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala (hunk 1 above)
Comment (agrawaldevesh) [reply core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala:1074]: Got it. Thanks for considering it.

File: core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (agrawaldevesh) [reply core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala:1065]: I am just a bit wary of regressions and I am not too sure about the unit test coverage here. So generally like the code changes to be minimal. But I am okay with this change if you think it is safe.

**Comment by agrawaldevesh** [comment]
Thanks for updating the PR description. Now all we need is a real test to validate that the PR is actually fixing the delay scheduling issue you pointed out. The current test is not really testing that but rather just testing the "definition of aliveness" -- so it is good to retain as well.

File: core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala [hunk 3]
```diff
@@ -1062,25 +1062,36 @@ private[spark] class TaskSchedulerImpl(
...
   }
 
   def hasExecutorsAliveOnHost(host: String): Boolean = synchronized {
-    hostToExecutors.contains(host)
+      hostToExecutors.get(host)
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala:1069]: wrong indentation

File: core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (cloud-fan) [reply core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala:1065]: It's super weird if `getExecutorsAliveOnHost` and `hasExecutorsAliveOnHost` are not consistent.

**Comment by Ngone51** [comment]
> Now all we need is a real test to validate that the PR is actually fixing the delay scheduling issue you pointed out. 

@agrawaldevesh I added a unit test for locality level computation in `TaskSetManagerSuite`. I'm not sure what do you mean by a "real test".  But if it means an end to end test which validates this fix can avoid the unnecessary delay, then I think it can be very hard.

**Review by cloud-fan — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
retest this please

**Review by agrawaldevesh — COMMENTED** [review]
Thanks @Ngone51 ... I verified that this is a indeed an issue in decommissioning right now: We do end up causing more delay scheduling delay and this PR indeed fixes that. This is a great find and an important hole to fix. Thanks for the careful reasoning and insight that went behind it.

My main blocker here is about the `TaskSchedulerImpl.isExecutorDecommissioned` method added. I would please urge you to consider my suggestion. I think we can also add more testing for the case when removing just a single executor. 

To ease your life, I tested all these changes and added extra tests. Please consider incorporating the commit https://github.com/agrawaldevesh/spark/commit/bcd197769eb46d9dee637c2942fc3a5aadd32424. 
 Please let me know if you cannot access this commit -- I couldn't really create a PR on your branch because your repo didn't show up for me.

File: core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala [hunk 4]
```diff
@@ -1062,25 +1062,38 @@ private[spark] class TaskSchedulerImpl(
...
+    executorIdToRunningTaskIds.contains(execId) && !isExecutorDecommissioned(execId)
   }
 
   def isExecutorBusy(execId: String): Boolean = synchronized {
     executorIdToRunningTaskIds.get(execId).exists(_.nonEmpty)
   }
 
+  // exposed for test
+  protected def isExecutorDecommissioned(execId: String): Boolean =
+    getExecutorDecommissionInfo(execId).nonEmpty || isHostDecommissioned(executorIdToHost(execId))
```
[the remark this answers is not in the document]
Comment (Ngone51) [reply core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala]: make sense to me.

**Comment by Ngone51** [comment]
@agrawaldevesh Thanks for your commit. Changes make sense to me. I've merged it into this PR.

**Review by agrawaldevesh — APPROVED** [review]
Yay ! Cool catch: Really subtle and really good code review of the decommissioning logic. Thanks for fixing this ! This should help :-)

**Comment by cloud-fan** [comment]
retest this please

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by Ngone51** [comment]
thanks all!

**Pull request closed without merging**